plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzBatchService.groovy

```diff
@@ -252,16 +253,32 @@ class AzBatchService implements Closeable {
 
     protected BatchClient createBatchClient() {
         log.debug "[AZURE BATCH] Executor options=${config.batch()}"
-        // Create batch client
-        if( !config.batch().endpoint )
-            throw new IllegalArgumentException("Missing Azure Batch endpoint -- Specify it in the nextflow.config file using the setting 'azure.batch.endpoint'")
-        if( !config.batch().accountName )
-            throw new IllegalArgumentException("Missing Azure Batch account name -- Specify it in the nextflow.config file using the setting 'azure.batch.accountName'")
-        if( !config.batch().accountKey )
-            throw new IllegalArgumentException("Missing Azure Batch account key -- Specify it in the nextflow.config file using the setting 'azure.batch.accountKet'")
-
-        final cred = new BatchSharedKeyCredentials(config.batch().endpoint, config.batch().accountName, config.batch().accountKey)
-        final client = BatchClient.open(cred)
+
+        def cred
+
+        if (config.batch().endpoint || config.batch().accountKey || config.batch().accountName ) {
+             // Create batch client
+            if( !config.batch().endpoint )
+                throw new IllegalArgumentException("Missing Azure Batch endpoint -- Specify it in the nextflow.config file using the setting 'azure.batch.endpoint'")
+            if( !config.batch().accountName )
+                throw new IllegalArgumentException("Missing Azure Batch account name -- Specify it in the nextflow.config file using the setting 'azure.batch.accountName'")
+            if( !config.batch().accountKey )
+                throw new IllegalArgumentException("Missing Azure Batch account key -- Specify it in the nextflow.config file using the setting 'azure.batch.accountKet'")
+
+            cred = new BatchSharedKeyCredentials(config.batch().endpoint, config.batch().accountName, config.batch().accountKey)
+        } else {
+
+            final batchEndpoint = config.batch().batchEndpoint ?: "https://batch.core.windows.net/";
+            final authenticationEndpoint =  config.batch().authenticationEndpoint ?: "https://login.microsoftonline.com/";
+            final batchServiceEndpoint = config.batch().batchEndpoint
+            final tenantId = config.batch().tenantId
+            final clientId = config.batch().clientId
+            final clientSecret = config.batch().clientSecret
+            cred = new BatchApplicationTokenCredentials(batchServiceEndpoint, clientId, clientSecret, tenantId, batchEndpoint, authenticationEndpoint)
+        }
```

Comment:
Ohh, that's it?

Reply:
On the code side yeah, the changes won't be a lot, but a lot of time was spent on 
1. Getting the batch/pool level configs right for testing
2. Due to Azure quota issues (batch accounts), I can only iterate on a single account 

All this lead to a rabbit hole for terraform templates to bootstrap the testing env.


But not done yet sorry - the compiled code needs testing and improvements :sweat_smile:

Reply:
Cool, well done